test_util/cluster.py

```diff
@@ -0,0 +1,292 @@
+        )
+
+        return cls(
+            ssh_info=ssh_info,
+            ssh_key_path=ssh_key_path,
+            provider='onprem',
+            masters=masters,
+            agents=agents,
+            public_agents=public_agents,
+            bootstrap_host=bootstrap_host,
+            cleanup_callback=cleanup_callback,
+        )
+
+    @classmethod
+    def from_vpc(
+        cls,
+        vpc,
+        ssh_info,
+        ssh_key_path,
+        num_masters,
+        num_agents,
+        num_public_agents,
+    ):
+        hosts = vpc.get_vpc_host_ips()
+        logging.info('AWS provided VPC info: ' + repr(hosts))
+
+        return cls.from_hosts(
+            ssh_info=ssh_info,
+            ssh_key_path=ssh_key_path,
+            hosts=hosts,
+            num_masters=num_masters,
+            num_agents=num_agents,
+            num_public_agents=num_public_agents,
+            cleanup_callback=vpc.delete,
+        )
+
+    @classmethod
+    def from_cloudformation(cls, cf, ssh_info, ssh_key_path):
+        return cls(
+            ssh_info=ssh_info,
+            ssh_key_path=ssh_key_path,
+            provider='aws',
+            masters=cf.get_master_ips(),
+            agents=cf.get_private_agent_ips(),
+            public_agents=cf.get_public_agent_ips(),
+            cleanup_callback=cf.delete,
+        )
+
+    def cleanup(self):
+        if self.cleanup_callback is not None:
+            self.cleanup_callback()
+
+    @property
+    def hosts(self):
+        return self.masters + self.agents + self.public_agents + (
+            [self.bootstrap_host] if self.bootstrap_host else []
+        )
+
+    @staticmethod
+    def partition_cluster(hosts, num_masters, num_agents, num_public_agents):
+        """Return (bootstrap, masters, agents, public_agents) from hosts."""
+        hosts_iter = iter(hosts)
+        return (
+            next(hosts_iter),
+            list(itertools.islice(hosts_iter, num_masters)),
+            list(itertools.islice(hosts_iter, num_agents)),
+            list(itertools.islice(hosts_iter, num_public_agents)),
+        )
+
+    @retry(stop_max_delay=120000)
+    def check_ssh(self, hosts):
+        """SSH to each host's public IP and run a command.
+
+        Retries on failure for up to 2 minutes.
+
+        """
+        self.ssher.remote_cmd([h.public_ip for h in hosts], ['echo'])
+
+
+def install_dcos(
+    cluster,
+    installer_url,
+    api=False,
+    setup=True,
+    add_config_path=None,
+    installer_api_offline_mode=True,
+    install_prereqs=True,
+    install_prereqs_only=False,
+):
+    if cluster.provider != 'onprem':
+        raise NotImplementedError('Install is only supported for onprem clusters')
+
+    assert cluster.bootstrap_host is not None, 'Install requires a bootstrap host'
+    assert all(h.public_ip for h in cluster.hosts), 'Install requires that all cluster hosts are externally reachable'
+
+    if api:
+        installer = test_util.installer_api_test.DcosApiInstaller()
```

Comment:
nit / not yours as I'm passing by this. Bot the web installer and ssh installer have the same "offline". Web it's called "offline mode", ssh it's called "install_prereqs=False". We should clean that up at some point (not this pr)